lib/network/level-net/sio-stream.js

```diff
@@ -0,0 +1,372 @@
+        this.remoteStreams = {};
+        this.localStreams = {};
+
+        this.socket.on('stream-data', (payload, cb) => {
+            const { streamId, data } = payload;
+            debug('received stream-data event streamId', streamId,
+                  'data', data);
+            const stream = this.remoteStreams[streamId];
+            if (!stream) {
+                debug('no remote stream registered with id', streamId);
+                return;
+            }
+            stream._ondata(data);
+            cb(null);
+        });
+
+        this.socket.on('stream-end', (payload, cb) => {
+            const { streamId } = payload;
+            debug('received stream-end event streamId', streamId);
+            const stream = this.remoteStreams[streamId];
+            if (!stream) {
+                debug('no remote stream registered with id', streamId);
+                return;
+            }
+            stream._onend();
+            cb(null);
+        });
+
+        this.socket.on('stream-error', payload => {
+            const { streamId, error } = payload;
+            debug('received stream-error event streamId', streamId,
+                 'error', error);
+            const stream = this.remoteStreams[streamId];
+            if (!stream) {
+                debug('no remote stream registered with id', streamId);
+                return;
+            }
+            stream._onerror(error);
+            delete this.remoteStreams[streamId];
+        });
+
+        this.socket.on('stream-hangup', payload => {
+            const { streamId } = payload;
+            debug('received stream-hangup event streamid', streamId);
+            const stream = this.localStreams[streamId];
+            if (!stream) {
+                debug('no local stream registered with id', streamId,
+                      '(may have already reached the end)');
+                return;
+            }
+            this.destroyStream(streamId);
+        });
+    }
+
+    encodeStreams(arg) {
+        if (!arg) {
+            return arg;
+        }
+        const isReadStream = (typeof(arg.pipe) === 'function'
+                              && typeof (arg.read) === 'function');
+        const isWriteStream = (typeof(arg.write) === 'function');
+
+        if (isReadStream || isWriteStream) {
+            const streamId = uuid();
+            const encodedStream = {
+                $streamId: streamId,
+                readable: isReadStream,
+                writable: isWriteStream,
+            };
+            debug('encoded stream:', encodedStream);
+            let transportStream;
+            if (isReadStream && isWriteStream) {
+                transportStream = new SIODuplexStream(this, streamId,
+                                                      this.maxPendingAck,
+                                                      this.ackTimeoutMs);
+            } else if (isReadStream) {
+                transportStream = new SIOOutputStream(this, streamId,
```

Comment:
I removed the duplex streams since a. they are not used and b. they are probably broken.

I think we can keep the support for both readable and writable streams though, as it only adds a few lines to encode and decode functions as the rest is fully symmetric (events can be emitted or listened to by either side the same way) and I can imagine use cases where writable streams on the client could be useful (possibly write gathering) otherwise I think they do no harm being unused (but you may disagree).